### PR TITLE
gameover UI

### LINKED ISSUE #9: source_remove warnings
```log
/usr/share/sugar/activities/TurtleinaPond.activity/game.py:167: Warning: Source ID 12324 was not found when attempting to remove it
  GObject.source_remove(self._timeout_id)
```

### PR DESCRIPTION
**Pull request by sourabhaa**
This PR adds a game-over page to the activity.
Fixes #9

### CODE REVIEW

**Comment by sourabhaa** [comment]
@chimosky @srevinsaju @JuiP @Saumya-Mishra9129 kindly review.

File: game.py [hunk 1]
```diff
@@ -256,9 +268,76 @@ def _test_game_over(self, new_dot):
            # Game-over feedback
            for dot in self._dots:
                dot.set_label(':)')
+           self._timeout_id = GLib.timeout_add(4000, self._game_over)
            return True
-        return False
 
+        return False
```
Comment (chimosky) [inline game.py:293]: You should leave an empty line before a new function definition.

File: game.py [hunk 2]
```diff
@@ -382,6 +462,17 @@ def do_expose_event(self, event):
...
+        ''' generate a dot of a color color '''
+        self._stroke = color
+        self._fill = color
+        self._svg_width = self._dot_size_gameover
+        self._svg_height = self._dot_size_gameover
+        return svg_str_to_pixbuf(
+            self._header() + \
+            self._circle(self._dot_size_gameover / 2., self._dot_size_gameover / 2.,
+                         self._dot_size_gameover / 2.) + \
+            self._footer())
```
Comment (chimosky) [inline game.py]: Would be more readable if the operator is before the next operand and it's indent should be same as the other operand.

**Review by chimosky — CHANGES_REQUESTED** [review]

File: game.py [hunk 3]
```diff
@@ -12,6 +12,7 @@
 # along with this library; if not, write to the Free Software
 # Foundation, 51 Franklin Street, Suite 500 Boston, MA 02110-1335 USA
 
+# from TurtlePondActivity import TurtlePondActivity
```
Comment (chimosky) [inline game.py]: You can delete this line if it was just for a test.

File: game.py [hunk 4]
```diff
@@ -284,9 +287,9 @@ def _game_over(self):
         for x in range(2, 6):
             self._gameover.append(
                 Sprite(self._sprites,
-                       xoffset + (x - 0.5) * (self._dot_size + 50),
-                       y * (self._dot_size + 200 + self._space) + yoffset,
-                       self._new_dot_gameover('#A0FFA0')))
+                       xoffset + (x - 0.5) * (self._dot_size + 140),
+                       y * (self._dot_size + 140 + self._space) + yoffset,
+                       self._new_dot_gameover(self._colors[FILL])))
```
Comment (chimosky) [inline game.py]: Might be better if these values are stored as a constant as they're reused later.

**Review by chimosky — CHANGES_REQUESTED** [review]

File: game.py [hunk 5]
```diff
@@ -190,10 +198,23 @@ def new_game(self, saved_state=None):
             if self._dots[n].type is not None:
                 self._dots[n].type = True
                 self._dots[n].set_shape(self._new_dot(self._colors[STROKE]))
-
         # Calculate the distances to the edge
         self._initialize_weights()
         self.strategy = self.strategies[self.level]
+        self.thread = threading.Thread(target = self.timer, daemon = True)
```
Comment (chimosky) [inline game.py]: Keyword args shouldn't be spaced, also why introduce threads?

File: game.py [hunk 6]
```diff
@@ -190,10 +198,23 @@ def new_game(self, saved_state=None):
...
+
+    def timer(self):
+        self.time_now = 0
+        while (self.gameover_flag is False) and (self.time_now < 180):
+            time.sleep(1)
+            self.time_now += 1
+            self.seconds += 1
+            if self.seconds >= 60:
+                self.minutes = self.seconds // 60
+                self.minutes = self.minutes % 60
```
Comment (chimosky) [inline game.py]: This can be on one line, you can use parentheses to separate both operations.

File: game.py [hunk 7]
```diff
@@ -299,12 +304,12 @@ def _test_game_over(self, new_dot):
         return False
 
     def _game_over(self):
-        best_time = 0
-        best_seconds = 0
-        best_minutes = 0
+        best_seconds = self.best_time % 60
+        best_minutes = self.best_time // 60
```
Comment (chimosky) [inline game.py:297]: This can be on one line too.

**Comment by chimosky** [comment]
Thanks, reviewed 7cf5beb, not tested.

File: game.py (hunk 5 above)
Comment (sourabhaa) [reply game.py]: > also why introduce threads

Since there is no game loop, I had to use threads to run the stopwatch independently. I'm not sure if there is a way to avoid using it.

File: game.py (hunk 7 above)
Comment (sourabhaa) [reply game.py:297]: these are different variables, not sure if I can merge it.

File: game.py [hunk 8]
```diff
@@ -12,12 +12,18 @@
...
+# from TurtlePondActivity import TurtlePondActivity
 from gi.repository import Gtk, Gdk, GdkPixbuf, GLib
 
 import cairo
 
 from math import sqrt, pi
 from random import uniform
+import threading
+import time
+import os
```
Comment (srevinsaju) [inline game.py]: Reorder imports, imports should be sorter preferably by alphabetical order.

The order of imports, for flake should be:
1. first plain imports. eg: `import cairo`
2. from imports. eg: `from random import uniform`

So you should move L22-L24 before import to cairo.

File: game.py [hunk 9]
```diff
@@ -256,9 +301,121 @@ def _test_game_over(self, new_dot):
...
         return False
 
+    def _game_over(self):
+        best_seconds = self.best_time % 60
+        best_minutes = self.best_time // 60
+        for dot in self._dots:
+            dot.hide()
+        self._turtle.hide()
+        self.save_best_time()
+        yoffset = int(self._space_gameover / 4.)
```
Comment (srevinsaju) [inline game.py]: I think it is better to use `y_offset` instead of `yoffset`. Maybe `offset_y` is better? What do you think?

File: game.py [hunk 10]
```diff
@@ -292,6 +450,7 @@ def _happy_turtle_dance(self):
         self._orientation %= 6
         self._turtle.set_shape(self._turtle_images[self._orientation])
         self._timeout_id = GLib.timeout_add(250, self._happy_turtle_dance)
+        return
```
Comment (srevinsaju) [inline game.py]: Not really sure. Is this empty `return` statement required?

File: game.py [hunk 11]
```diff
@@ -382,6 +540,17 @@ def do_expose_event(self, event):
     def _destroy_cb(self, win, event):
         Gtk.main_quit()
 
+    def _new_dot_gameover(self, color):
+        ''' generate a dot of a color color '''
```
Comment (srevinsaju) [inline game.py:535]: The docstring at L555, and L544 matches. Can you add further explanation on how `_new_dot_gameover` is different `_new_dot`

File: game.py [hunk 12]
```diff
@@ -547,10 +529,11 @@ def _new_dot_gameover(self, color):
...
         return svg_str_to_pixbuf(
-            self._header() + \
-            self._circle(self._dot_size_gameover / 2., self._dot_size_gameover / 2.,
-                         self._dot_size_gameover / 2.) + \
-            self._footer())
+                            self._header()
+                            + self._circle(self._dot_size_gameover / 2.,
+                                           self._dot_size_gameover / 2.,
+                                           self._dot_size_gameover / 2.)
+                            + self._footer())
```
Comment (chimosky) [inline game.py]: I like the change here but can you keep the old indentation, I don't see a reason for increasing it this way.

File: game.py (hunk 5 above)
Comment (chimosky) [reply game.py]: I understand now, thanks.

File: game.py (hunk 7 above)
Comment (chimosky) [reply game.py:297]: Yeah didn't notice that, thanks.

File: game.py (hunk 12 above)
Comment (sourabhaa) [reply game.py]: if I just placed the operator before the next operand keeping the old indentation, line 533 would cross 79 character per line limit and the next line would be over-indented. Hence I had to change the indentation this way.

Comment (chimosky) [reply game.py]: I don't see how that'll be the case as keeping the old indentation would look something like this;
```python3
def _new_dot_gameover(self, color):
    ''' generate a dot of a color color '''
    self._stroke = color
    self._fill = color
    self._svg_width = self._dot_size_gameover
    self._svg_height = self._dot_size_gameover
    return svg_str_to_pixbuf(
        self._header()
        + self._circle(self._dot_size_gameover / 2.,
                       self._dot_size_gameover / 2.,
                       self._dot_size_gameover / 2.)
        + self._footer())
```

Comment (sourabhaa) [reply game.py]: Oh my bad, I had over indented line 534 and 535 when I tried keeping the old indentation n tested for flake errors, so I'd got over indentation error previously😅

**Comment by chimosky** [comment]
Tested 408fc03, the game over page overlaps to the lower part of the screen.

Screenshot from 2021-08-06 02-36-47

**Comment by sourabhaa** [comment]
> Tested 408fc03, the game over page overlaps to the lower part of the screen.

@chimosky, I fixed this. But I'm surprised to see the colour of the circles on your screen. I used the same colour as in the pond, not sure how it appears differently on your system🤔. But even this looks good though ;)

File: game.py [hunk 13]
```diff
@@ -382,6 +462,17 @@ def do_expose_event(self, event):
     def _destroy_cb(self, win, event):
         Gtk.main_quit()
 
+    def _new_dot_gameover(self, color):
+        ''' generate a dot of a color color '''
```
Comment (chimosky) [inline game.py]: There's a repetition of color at the end.

**Comment by chimosky** [comment]
> @chimosky, I fixed this. But I'm surprised to see the colour of the circles on your screen. I used the same colour as in the pond, not sure how it appears differently on your systemthinking. But even this looks good though ;)

Same colour as in pond? The colours shown are from my user profile and from [`game.py`](https://github.com/sugarlabs/turtlepond/blob/master/TurtlePondActivity.py#L77) the colours passed are from the user profile, it'll be different on your screen and not the colours you chose either as your user profile is different from mine.
There's also a repetition of assigning values to the kwarg `colors` of `Game`, you could set the param to `None` so as to stop the repetition.

You can check your user profile colours like this - this is mine -
```python3
>>> from sugar3 import profile
>>> c = profile.get_color()
>>> c.to_string()
'#00EA11,#BE9E00'
```

Could you also centralize the gameover page as it isn't?
Screenshot from 2021-08-06 15-56-36

File: game.py (hunk 13 above)
Comment (chimosky) [reply game.py]: @sourabhaa was this intended?

**Comment by chimosky** [comment]
Tested 3aac057, got a `Source ID not found` warning but I don't know what action triggered it and I also noticed that when a game ends and a new game is triggered before the game over page shows, the game over page eventually shows a couple of seconds later - seems like it's the number of seconds it'll take the turtle to go round the pond - and the values shown are all negative. In one of the cases after losing a game, the gameover page showed me that I'd won - screenshot below - and the time seems to be constant.

```
/home/ibiam/Activities/turtlepond/game.py:190: Warning: Source ID 1940 was not found when attempting to remove it
  GLib.source_remove(self._timeout_id)
```


Screenshot from 2021-08-07 15-45-41

**Comment by sourabhaa** [comment]
> got a Source ID not found warning but I don't know what action triggered it

I noticed that, since it was not introduced from this PR, I thought of taking a look at it after all the bugs from this PR is fixed.

>  I also noticed that when a game ends and a new game is triggered before the game over page shows, the game over page eventually shows a couple of seconds later - seems like it's the number of seconds it'll take the turtle to go round the pond

The gameover page shows up after the turtle goes round the pond. The new game is triggered after the gameover page shows up.

> and the values shown are all negative.  In one of the cases after losing a game, the gameover page showed me that I'd won - screenshot below - and the time seems to be constant.

Woah! 🙄 This never happened on my system, I'm testing in all possible ways. Could you please tell how exactly you played to get this result? I'm not able to reproduce it.

**Comment by chimosky** [comment]
> The gameover page shows up after the turtle goes round the pond. The new game is triggered after the gameover page shows up.

Yes that's the case but my test was to trigger a new game before the turtle goes round the pond, one way to fix this is to set the sensitivity on the new game button to be false right when a game ends so the turtle can go round the pond and show the game over page then the button sensitivity can change after that.

>Woah! 🙄 This never happened on my system, I'm testing in all possible ways. Could you please tell how exactly you played to get this result? I'm not able to reproduce it.

Sure, steps to reproduce;
* Start a game.
* Play and lose so the turtle would go round the pond.
* While the turtle is going round the pond, click the `Start new game` button so the turtle doesn't completely go round the pond.
* New game starts but some seconds later the game over page shows and then you can see those values.

**Comment by chimosky** [comment]
Tested 1fe0e7b, works as expected. Thanks.
Fixed source remove warning in 5eefcb1, you can test it.

I made a comment above about the docstring in `_new_dot_gameover`, I'm guessing you didn't see it.

File: game.py (hunk 13 above)
Comment (sourabhaa) [reply game.py]: @chimosky sorry, I missed it earlier. Yes, the former refers to the noun color, latter refers to the argument `color`.

**Comment by sourabhaa** [comment]
> Fixed source remove warning in 5eefcb1, you can test it.

Sure, tested 5eefcb1, works as expected :)

File: game.py [hunk 14]
```diff
@@ -684,11 +684,12 @@ def save_best_time(self):
...
-            try:
-                with open(file_path, "r") as fp:
-                    best_time = fp.readlines()
-                return int(best_time[0])
-            except(ValueError, IndexError):
+            with open(file_path, "r") as fp:
+                highscore = fp.readlines()
+            try: 
+                return int(highscore[0])
+            except  Exception as e:
```
Comment (chimosky) [inline game.py]: Extra spaces before `Exception` and why the change from `(ValueError, IndexError)`?

**Review by chimosky — CHANGES_REQUESTED** [review]

**Comment by chimosky** [comment]
Also for future reference, could you avoid merge commits - 0a63ed7 - as it makes viewing git history with a diff tool a bit difficult, you could've easily pulled my change to your local branch and then commit on top of that.

**Review by srevinsaju — CHANGES_REQUESTED** [review]
Ouch, I reviewed this last week, but I forgot to hit "submit review" smh :facepalm:

File: game.py [hunk 15]
```diff
@@ -18,6 +18,11 @@
 
 from math import sqrt, pi
 from random import uniform
+import threading
```
Comment (srevinsaju) [inline game.py]: Not sure, but... I can't see any references of threading module in the code, are we still using it?

**Review by srevinsaju — CHANGES_REQUESTED** [review]

File: game.py (hunk 8 above)
Comment (sourabhaa) [reply game.py]: > So you should move L22-L24 before import to cairo.

Shouldn't it be after import to cairo?🤔

File: game.py (hunk 9 above)
Comment (sourabhaa) [reply game.py]: Yeah, that's more readable. In that case I'll change `xoffset` too.

File: game.py (hunk 11 above)
Comment (sourabhaa) [reply game.py:535]: The dimensions are different. Maybe I could refactor it.

File: game.py (hunk 15 above)
Comment (sourabhaa) [reply game.py]: Oh yeah, I forgot to remove it when I discarded the previous approach. I'll remove it.

File: game.py (hunk 10 above)
Comment (sourabhaa) [reply game.py]: I guess it's not needed, I don't remember adding that🤔, maybe I put it there by mistake

**Comment by chimosky** [comment]
Reviewed, thanks. Merged by hand.
